From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch 03-03
Patch-Mainline: 2.6.18-rc1
References: 161997 - LTC22671

  Description: kernel: software watchdog crashes.
  Symptom:     If the software watchdog timer expires the machine panics
               instead of doing a restart.
  Problem:     The software watchdog calls machine_restart from its timer
               interrupt function. machine_restart calls console_unblank
               which checks for interrupt context and triggers a BUG()
               if the code is in interrupt context.
  Solution:    Avoid the call to console_unblank in machine_restart if
               the function has been called in interrupt context and
               oops_in_progress is not set.
  Problem-ID:  22671
  Archs:       s390-31, s390-64

Acked-by: Andi Kleen <ak@suse.de>
Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/arch/s390/kernel/setup.c	20 Mar 2006 15:56:34 -0000	1.71
+++ linux-2.5/arch/s390/kernel/setup.c	5 May 2006 16:53:33 -0000	1.71.2.1
@@ -297,19 +297,34 @@
 
 void machine_restart(char *command)
 {
-	console_unblank();
+	if (!in_interrupt() || oops_in_progress)
+		/*
+		 * Only unblank the console if we are called in enabled
+		 * context or a bust_spinlocks cleared the way for us.
+		 */
+		console_unblank();
 	_machine_restart(command);
 }
 
 void machine_halt(void)
 {
-	console_unblank();
+	if (!in_interrupt() || oops_in_progress)
+		/*
+		 * Only unblank the console if we are called in enabled
+		 * context or a bust_spinlocks cleared the way for us.
+		 */
+		console_unblank();
 	_machine_halt();
 }
 
 void machine_power_off(void)
 {
-	console_unblank();
+	if (!in_interrupt() || oops_in_progress)
+		/*
+		 * Only unblank the console if we are called in enabled
+		 * context or a bust_spinlocks cleared the way for us.
+		 */
+		console_unblank();
 	_machine_power_off();
 }
 

